fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/item/v1/RecipeRemainderProvider.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.item.v1;
+
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.inventory.Inventory;
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemStack;
+import net.minecraft.recipe.RecipeType;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+
+/**
+ * Allows an item to conditionally specify the recipe remainder.
+ * The recipe remainder is an {@link ItemStack} instead of an {@link Item}.
+ * This can be used to allow your item to get damaged instead of
+ * getting removed when used in crafting.
+ *
+ * <p>Recipe remainder providers can be set with {@link FabricItemSettings#recipeRemainder(RecipeRemainderProvider)}</p>
+ */
+@FunctionalInterface
+public interface RecipeRemainderProvider {
```

Comment:
Is this exposed somewhere for custom crafting devices like modded crafting tables or processing machines?

Reply:
yes, provided that the modded crafting table calls `RecipeManager#getRemainingStacks` to get recipe remainders

Reply:
should probably add a method to get the recipe remainder of an ItemStack